### PR TITLE
feat: add gettext to html and javascript

### PR DESCRIPTION
**Pull request by OmarIthawi**
Replace hardcoded strings with gettext calls.

References
----------

This pull request is part of the [FC-0012 project](https://openedx.atlassian.net/l/cp/XGS0iCcQ) which is sparked by the [Translation Infrastructure update OEP-58](https://open-edx-proposals.readthedocs.io/en/latest/architectural-decisions/oep-0058-arch-translations-management.html#specification).

Check the links above for full information about the overall project.

Internalization is being rearchitected in Open edX Python, XBlock, Micro-frontend, and other projects. There are a number of immediately visible changes:
 - Remove source and language translations from the repositories, hence no `.json`, `.po` or `.mo` files will be committed into the repos.
 - Add standardized `make extract_translations` in all repositories
 - Push user-facing messages strings into [openedx/openedx-translations](https://github.com/openedx/openedx-translations/).
 - Integrate root repositories with [openedx/openedx-atlas](https://github.com/openedx/openedx-atlas/) to pull translations on build/deploy time

Breaking Changes
----------------

One of the primary goals of the project is to **avoid breaking changes**. If you noticed any suspicious code, please raise your concern. But before that, please know the strategy we're following to avoid breaking changes

**For XBlocks:**
 - The standard translation path must be `conf/locale`. Therefore, we are creating a link from `conf/locale` to `translations` so Atlas can find the path without disrupting the current way of having translations locally within the XBlocks
 - [openedx-translations](https://github.com/openedx/openedx-translations) will have a related PR that adds the XBlock to the pipeline. This will not affect the current way of managing/updating translations
 - At the end of the project, a clear change log will be added and all translations will be handled by Atlas. Thus, the local translation will be removed from the repo within the version bump
 - A notification for the community will be published to ensure that everyone knows why translations directories are removed from all repos

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @OmarIthawi!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

**Comment by shadinaif** [comment]
Awsome @OmarIthawi ! I've tested this on my local tutor and it rendered fine 😄

File: recommender/static/html/recommender.html [hunk 1]
```diff
@@ -76,18 +78,15 @@
 <div class="recommenderBlock">
     <div class="recommenderRecommendations">
         <div class="recommenderContent">
-            <div class="recommender_hideShow" role='button' tabindex='0' data-step="1" aria-expanded='false' data-intro="This is a list of recommended resources. If you're stuck, curious, or want to learn more, please look around. If you find a useful resources, help your fellow classmates by adding it!">
+            <div class="recommender_hideShow" role='button' tabindex='0' data-step="1" aria-expanded='false' data-intro="${_("This is a list of recommended resources. If you're stuck, curious, or want to learn more, please look around. If you find a useful resources, help your fellow classmates by adding it!")}">
                 Show related resources
```
Comment (shadinaif) [inline recommender/static/html/recommender.html:82]: @OmarIthawi is there a reason we're not translating some texts like this one? or is it just missed?

**Review by shadinaif — APPROVED** [review]

File: recommender/static/html/recommender.html (hunk 1 above)
Comment (OmarIthawi) [reply recommender/static/html/recommender.html:82]: I missed it.

Comment (shadinaif) [reply recommender/static/html/recommender.html]: I've added them all @OmarIthawi 👍🏼

**Comment by OmarIthawi** [comment]
@mphilbrick211 @brian-smith-tcril this now ready for review and merge.

File: recommender/static/html/recommender.html (hunk 1 above)
Comment (OmarIthawi) [reply recommender/static/html/recommender.html]: Thanks for the help :)

File: recommender/static/html/recommender.html [hunk 2]
```diff
@@ -129,109 +128,102 @@
...
-                        placeholder="http://en.wikipedia.org/wiki/Edx"/>
+                        placeholder="http://en.wikipedia.org/wiki/edX"/>
                 </label>
-                <label>Description&nbsp;
+                <label>${_("Description")}&nbsp;
                     <textarea
                         aria-label='Provide a resource description'
                         class="recommender_editDescriptionText"
-                        placeholder="Provide a meaningful description so other students know whether this is useful to them"></textarea>
+                        placeholder="${_("Provide a meaningful description so other students know whether this is useful to them")}"></textarea>
```
Comment (shadinaif) [inline recommender/static/html/recommender.html:197]: @OmarIthawi it renders fine, but I found that the translation is broken due to double-quotation sequence I guess, I'll continue testing and be back with a fix

**Review by shadinaif — CHANGES_REQUESTED** [review]

File: recommender/static/html/recommender.html (hunk 2 above)
Comment (shadinaif) [reply recommender/static/html/recommender.html:197]: sorry, my bad. Broken in my tutor. Looks like I need some caffeine  😅

**Review by shadinaif — APPROVED** [review]

File: recommender/static/html/recommender.html (hunk 2 above)
Comment (shadinaif) [reply recommender/static/html/recommender.html]: I've pushed the double-qutation fix, @OmarIthawi

Comment (OmarIthawi) [reply recommender/static/html/recommender.html]: Thanks! done.

File: recommender/static/html/recommender.html [hunk 3]
```diff
@@ -76,18 +78,15 @@
 <div class="recommenderBlock">
     <div class="recommenderRecommendations">
         <div class="recommenderContent">
-            <div class="recommender_hideShow" role='button' tabindex='0' data-step="1" aria-expanded='false' data-intro="This is a list of recommended resources. If you're stuck, curious, or want to learn more, please look around. If you find a useful resources, help your fellow classmates by adding it!">
-                Show related resources
+            <div class="recommender_hideShow" role='button' tabindex='0' data-step="1" aria-expanded='false' data-intro="${_('This is a list of recommended resources. If you are stuck, curious, or want to learn more, please look around. If you find a useful resources, help your fellow classmates by adding it!')}">
```
Comment (brian-smith-tcril) [inline recommender/static/html/recommender.html:81]: I don't want to assume too much, but is the switch from `you're` to `you are` just because `gettext` is using single quoted strings?

If so, this should probably be
```suggestion
            <div class="recommender_hideShow" role='button' tabindex='0' data-step="1" aria-expanded='false' data-intro="${_('This is a list of recommended resources. If you\'re stuck, curious, or want to learn more, please look around. If you find a useful resources, help your fellow classmates by adding it!')}">
```

File: recommender/static/html/recommender.html [hunk 4]
```diff
@@ -116,11 +115,11 @@
                         <%include file="resourcebox.html" args="id=elem['id'],title=elem['title'],votes=elem['votes'],url=elem['url'],description=elem['description'],descriptionText=elem['descriptionText'],isProblematic=flag_mode,problematicReason=reason,voteMode=vote_mode,endorseMode=endorse_mode,endorsedReason=endorsed_reason" />
                     % endfor
                 </div>
-                <div class="recommender_resourceAddButton" role="button" tabindex="0">Add new resource &gt;&gt;</div>
-                <div class="recommender_resourceExportButton" role="button" tabindex="0">Download resources</div>
-                <div class="recommender_resourceImportButton" role="button" style='display: none;' tabindex="0" aria-hidden="true">Upload resources</div>
+                <div class="recommender_resourceAddButton" role="button" tabindex="0">${_("Add new resource")} &gt;&gt;</div>
```
Comment (brian-smith-tcril) [inline recommender/static/html/recommender.html:118]: should the `&gt;` be a part of the `gettext` string? It seems having it hardcoded after the `gettext` could be problematic with RTL languages

File: recommender/static/html/recommender.html [hunk 5]
```diff
@@ -129,109 +128,102 @@
...
                         <img class="recommender_previewingImg" style='display: none;' src="" aria-hidden="true"/>
                     </div>
                 </div>
             </div>
         </div>
         <div class='recommenderModify' style='display: none;' tabindex="0" role='region' aria-hidden="true">
             <div class='recommenderModifyTitleContainer'>
-                <div class='recommender_backToViewButton' role='button' tabindex="0" aria-label='cancel and go back to resource list'>&lt; Related resources</div>
+                <div class='recommender_backToViewButton' role='button' tabindex="0" aria-label='${_("cancel and go back to resource list")}'>&lt; ${_("Related resources")}</div>
```
Comment (brian-smith-tcril) [inline recommender/static/html/recommender.html:139]: same as the comment on "Add new resource" about RTL languages. Should we have `&lt;` as part of the `gettext` string here?

File: recommender/static/js/src/cats.js
Comment (brian-smith-tcril) [inline recommender/static/js/src/cats.js:1]: It's not clear from the PR comments why the change from `var` to `window.` is happening here. A little comment explaining why this is happening could go a long way!

**Review by brian-smith-tcril — COMMENTED** [review]
Overall this is looking really great! Left a couple comments with some questions/suggestions.

File: recommender/static/html/recommender.html (hunk 4 above)
Comment (OmarIthawi) [reply recommender/static/html/recommender.html:118]:
```suggestion
                <div class="recommender_resourceAddButton" role="button" tabindex="0">${_("Add new resource &gt;&gt;")}</div>
```

Yeah, it's safe to include it.

File: recommender/static/html/recommender.html (hunk 5 above)
Comment (OmarIthawi) [reply recommender/static/html/recommender.html:139]:
```suggestion
                <div class='recommender_backToViewButton' role='button' tabindex="0" aria-label='${_("cancel and go back to resource list")}'>${_("&lt; Related resources")}</div>
```

File: recommender/static/js/src/cats.js [hunk 6]
```diff
@@ -1,187 +1,203 @@
...
-        'previousVoteClassName': 'recommender_downvoting'
-    },
-    'downvote': {
-        'buttonClassName': 'recommenderVoteArrowDown',
-        'eventName': 'arrowDown',
-        'serverEventName': 'recommender_downvote',
-        'voteClassName': 'recommender_downvoting',
-        'previousVoteClassName': 'recommender_upvoting'
+(function (window) {
+    var gettext = window.gettext || (function (string) {
```
Comment (OmarIthawi) [inline recommender/static/js/src/cats.js:2]:
```suggestion
   // Wrapping this file in an Immediately Invoked Function (IIAF) to ensure that helpers such as `gettext` and `span` aren't poluting the global scope.
   //
   // Note: The global `window.*` variables should be converted into local ones to avoid over-using global variables. This is a 2014-era legacy XBlock coding standard that should be refactored -- @OmarIthawi at Apr 4, 2023 
   //
    var gettext = window.gettext || (function (string) {
```

File: recommender/static/js/src/cats.js
Comment (OmarIthawi) [reply recommender/static/js/src/cats.js:1]: I've added an explaination, it has to do with `gettext` override that needs to happen within the function call.

Also, the whole xblock architecture makes me 😢, but it's an 2014 era code that needs to be refactored.

**Review by brian-smith-tcril — APPROVED** [review]
LGTM :shipit:

**Review by sarina — APPROVED** [review]

**Comment by openedx-webhooks** [comment]
@OmarIthawi 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/RecommenderXBlock&entry.867055334=https://github.com/openedx/RecommenderXBlock/pull/66&entry.1484655318=https://github.com/OmarIthawi&entry.752974735=2023-03-13+19:47&entry.1917517419=2023-04-05+20:21&entry.2133058324=Yes) so we can improve your experience in the future.
